ZenPacks/zenoss/OpenStackInfrastructure/model_mappers.py

```diff
@@ -0,0 +1,1045 @@
+            serverBackupEnabled=backup_schedule_enabled,
+            serverBackupDaily=backup_schedule_daily,
+            serverBackupWeekly=backup_schedule_weekly,
+            publicIps=list(public_ips),
+            privateIps=list(private_ips),
+            set_flavor=prepId('flavor-{0}'.format(flavor_id)),
+            set_tenant=prepId('tenant-{0}'.format(tenant_id)),
+            hostId=server.get('hostId', ''),
+            hostName=server.get('name', ''),
+            powerState=POWER_STATE_MAP.get(power_state),
+            taskState=task_state,
+            vmState=vm_state,
+            )
+
+        # Some Instances are created from pre-existing volumes
+        # This implies no image exists.
+        image_id = None
+        if 'imageId' in server:
+            image_id = server['imageId']
+        elif 'image' in server \
+                and isinstance(server['image'], dict) \
+                and 'id' in server['image']:
+            image_id = server['image']['id']
+
+        if image_id:
+            server_dict['set_image'] = prepId('image-{0}'.format(image_id))
+
+        servers.append(ObjectMap(
+            modname='ZenPacks.zenoss.OpenStackInfrastructure.Instance',
+            data=server_dict))
+
+    # Add hypervisor instance name to the existing server objectmaps.
+    server_hypervisor_instance_name = results.get('process_server_hypervisor_instance_name')
+    for om in servers:
+        if om.id in server_hypervisor_instance_name:
+            om.hypervisorInstanceName = server_hypervisor_instance_name[om.id]
+
+    return servers
+
+
+def map_services(device, results, hostmap):
+    services = []
+    zones = {}
+    host_orgComponent = defaultdict(lambda: results.get('process_region_id'))
+
+    # Find all hosts which have a nova service on them.
+    for service in results['services']:
+        if not service.get('id', None):
+            continue
+
+        host_id = service['host']
+        try:
+            hostname = hostmap.get_hostname_for_hostid(host_id)
+        except InvalidHostIdException:
+            LOG.error("An invalid Host ID: '%s' was provided.\n"
+                      "\tPlease examine zOpenStackHostMapToId "
+                      "and zOpenStackHostMapSame.", host_id)
+            continue
+        except Exception:
+            LOG.warning("An unknown error for Host ID: '%s' occurred", host_id)
+            continue
+
+        host_base_id = re.sub(r'^host-', '', host_id)
+
+        title = '{0}@{1} ({2})'.format(service.get('binary', ''),
+                                       hostname,
+                                       service.get('zone', ''))
+        service_id = prepId('service-{0}-{1}-{2}'.format(
+            service.get('binary', ''),
+            host_base_id,
+            service.get('zone', '')))
+        zone_id = prepId("zone-{0}".format(service.get('zone', '')))
+
+        host_orgComponent[host_id] = zone_id
+
+        zones.setdefault(zone_id, ObjectMap(
```

Comment:
probably should just strip the zone stuff into its own method.. (duplicating this loop)